// grey-message/Message/index.js
var MessageItem = /** @class */ (function () {
    /** 构造函数 */
    function MessageItem(type, method) {
        this.type = type;
        this.method = method;
    }
    return MessageItem;
}());
var Message = /** @class */ (function () {
    function Message(thisObj) {
        var _this = this;
        /** 注册项列表 */
        this.registerList = [];
        /**
         * 发送消息
         * @param event 消息
         */
        this.send = function (event) {
            _this.registerList.forEach(function (item) {
                if (event.type === item.type) {
                    event.target = _this.thisObj;
                    item.method(event);
                }
            });
        };
        /**
         * 注册消息
         * @param type 消息类型
         * @param method 注册的代理函数
         */
        this.on = function (type, method) {
            if (_this.getIndex(type, method) === -1) {
                _this.registerList.push(new MessageItem(type, method));
            }
        };
        /**
         * 注销消息
         * @param type 消息类型
         * @param method 注册的代理函数
         */
        this.off = function (type, method) {
            var index = _this.getIndex(type, method);
            if (index >= 0) {
                _this.registerList.splice(index, 1);
            }
        };
        /**
         * 获取注册列表的索引
         * @param type 消息类型
         * @param method 注册的代理函数
         */
        this.getIndex = function (type, method) {
            for (var i = 0; i < _this.registerList.length; i++) {
                if (_this.registerList[i].method === method && _this.registerList[i].type === type) {
                    return i;
                }
            }
            return -1;
        };
        if (thisObj) {
            this.thisObj = thisObj;
        }
        else {
            this.thisObj = this;
        }
    }
    Message.typeId = 0;
    /** 创建消息类型 */
    Message.createType = function (name) {
        return name + "@@" + Message.typeId++;
    };
    return Message;
}());
export default Message;
